### PR TITLE
More of Player

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/rts/money.cpp [hunk 1]
```diff
@@ -41,6 +41,27 @@ void Money::Deposit(unsigned int amount, bool play_sound)
     }
 }
 
+unsigned int Money::Withdraw(unsigned int amount, bool play_sound)
+{
+    if (amount > m_money) {
+        amount = m_money;
+    }
+
+    if (!amount) {
```
Comment (xezon) [inline src/game/common/rts/money.cpp]: == 0

File: src/game/common/system/upgrade.h [hunk 2]
```diff
@@ -14,6 +14,7 @@
  */
 #pragma once
 #include "always.h"
+#include "academystats.h"
```
Comment (xezon) [inline src/game/common/system/upgrade.h:17]: prefer forward declaration over include if possible for better compile times.

```c++
enum AcademyClassificationType;
```

File: src/game/common/rts/specialpower.h [hunk 3]
```diff
@@ -14,7 +14,9 @@
  */
 #pragma once
 #include "always.h"
+#include "academystats.h"
```
Comment (xezon) [inline src/game/common/rts/specialpower.h:17]: prefer forward declaration over include if possible for better compile times.

```c++
enum AcademyClassificationType;
```

File: src/game/common/rts/specialpower.h [hunk 4]
```diff
@@ -14,7 +14,9 @@
...
 #include "audioeventrts.h"
+#include "object.h"
```
Comment (xezon) [inline src/game/common/rts/specialpower.h:19]: prefer forward declaration over include if possible for better compile times.

```c++
enum SpecialPowerType;
```

File: src/game/logic/object/update/aiupdate.h [hunk 5]
```diff
@@ -31,6 +31,85 @@ class Path;
...
+    AICMD_FOLLOW_WAYPOINT_PATH_EXACT,
+    AICMD_FOLLOW_WAYPOINT_PATH_AS_TEAM_EXACT,
+    AICMD_MOVE_AWAY_FROM_UNIT,
+    AICMD_FOLLOW_WAYPATH_APPEND,
+    AICMD_UNK50,
+    AICMD_GUARD_TUNNEL_NETWORK,
+    AICMD_EVACUATE_INSTANTLY,
+    AICMD_EXIT_INSTANTLY,
+    AICMD_GUARD_RETALIATE,
+    AICMD_NUM_COMMANDS,
```
Comment (xezon) [inline src/game/logic/object/update/aiupdate.h:91]: Can put a blank line between and remove the comma from the definite last enum.

File: src/game/logic/system/gamelogic.cpp [hunk 6]
```diff
@@ -312,3 +313,10 @@ Object *GameLogic::Friend_Create_Object(ThingTemplate const *thing, BitFlags<OBJ
     return nullptr;
 #endif
 }
+
+bool GameLogic::Is_In_Single_Player_Game()
```
Comment (xezon) [inline src/game/logic/system/gamelogic.cpp:317]: Function can be const.

File: src/game/logic/object/object.cpp [hunk 7]
```diff
@@ -651,3 +652,83 @@ float Object::Get_Shroud_Clearing_Range() const
 #endif
     return range;
 }
+
+bool Object::Has_Any_Special_Power() const
+{
+    return m_specialPowers.Any();
+}
+
+bool Object::Has_Special_Power(SpecialPowerType spt) const
```
Comment (xezon) [inline src/game/logic/object/object.cpp:661]: type

File: src/game/logic/object/object.cpp [hunk 8]
```diff
@@ -651,3 +652,83 @@ float Object::Get_Shroud_Clearing_Range() const
...
+
+        if (power != nullptr) {
+            const SpecialPowerTemplate *tmplate = power->Get_Special_Power_Template();
+
+            if (tmplate != nullptr) {
+                if (tmplate->Get_Type() == type) {
+                    return power;
+                }
+
+                if (type == SPECIAL_INVALID) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:679]: This condition is definitely correct? It reads odd.

File: src/game/common/rts/player.cpp [hunk 9]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+};
+
+void Do_Find_Command_Center(Object *obj, void *data)
+{
+    FindStruct *f = static_cast<FindStruct *>(data);
+
+    if (obj != nullptr) {
+        if (f->obj == nullptr && obj->Is_KindOf(KINDOF_COMMANDCENTER)) {
+            if (obj->Get_Template()->Get_Default_Owning_Side() == f->player->Get_Side()
+                && !obj->Get_Status(OBJECT_STATUS_UNDER_CONSTRUCTION) && !obj->Get_Status(OBJECT_STATUS_SOLD)) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:2297]: `!obj->Is_Effectively_Dead()` missing?

File: src/game/common/rts/player.cpp [hunk 10]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+                if (!power->Is_Script_Only()) {
+                    f->obj = obj;
+                    f->frame = 0;
+                }
+            }
+        }
+    }
+}
+
+Object *Player::Find_Most_Ready_Shortcut_Special_Power_Of_Type(SpecialPowerType type)
```
Comment (xezon) [inline src/game/common/rts/player.cpp:2361]: No out param for frame?

File: src/game/common/rts/player.cpp [hunk 11]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+    return f.count;
+}
+
+void Player::On_Structure_Construction_Complete(Object *builder, Object *structure, bool add_score)
+{
+    g_theScriptEngine->Notify_Of_Object_Creation_Or_Destruction();
+    g_theAI->Get_Pathfinder()->Remove_Object_From_Pathfind_Map(structure);
+    g_theAI->Get_Pathfinder()->Add_Object_To_Pathfind_Map(structure);
+
+    if (!add_score) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:2522]: This condition is correct? I would have expected the inverse.

File: src/game/common/rts/player.cpp [hunk 12]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+
+void Player::Set_Units_Should_Hunt(bool should_hunt, CommandSourceType source)
+{
+    m_unitsShouldHunt = should_hunt;
+    Coord3D loc;
+    g_thePartitionManager->Get_Most_Valuable_Location(
+        Get_Player_Index(), PLAYER_RELATIONSHIP_FLAGS_ENEMIES, VOT_VALUE, &loc);
+
+    for (auto i = m_playerTeamPrototypes.begin(); i != m_playerTeamPrototypes.end(); i++) {
+        DLINK_ITERATOR<Team> iter = (*i)->Iterate_Team_Instance_List();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: team_it

File: src/game/common/rts/player.cpp [hunk 13]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+
+        while (!iter.Done()) {
+            Team *team = iter.Cur();
+
+            if (team != nullptr) {
+                DLINK_ITERATOR<Object> iter2 = team->Iterate_Team_Member_List();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: member_it

File: src/game/common/rts/player.cpp [hunk 14]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+        }
+
+        m_money.Withdraw(m_money.Get(), true);
+    }
+}
+
+void Player::Set_Objects_Enabled(Utf8String template_type_to_affect, bool enable)
+{
+    for (auto i = m_playerTeamPrototypes.begin(); i != m_playerTeamPrototypes.end(); i++) {
+        DLINK_ITERATOR<Team> iter = (*i)->Iterate_Team_Instance_List();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: team_it

File: src/game/common/rts/player.cpp [hunk 15]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+
+        while (!iter.Done()) {
+            Team *team = iter.Cur();
+
+            if (team != nullptr) {
+                DLINK_ITERATOR<Object> iter2 = team->Iterate_Team_Member_List();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: member_it

File: src/game/common/rts/player.cpp [hunk 16]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+
+                while (!iter2.Done()) {
+                    Object *obj = iter2.Cur();
+
+                    if (obj != nullptr) {
+                        if (obj->Get_Template()->Get_Name().Compare(template_type_to_affect) == 0) {
+                            obj->Set_Script_Status(STATUS_ENABLED, !enable);
```
Comment (xezon) [inline src/game/common/rts/player.cpp:2740]: !enable correct?

File: src/game/common/rts/player.cpp [hunk 17]
```diff
@@ -2271,3 +2275,526 @@ bool Player::Calc_Closest_Construction_Zone_Location(const ThingTemplate *tmplat
...
+            iter.Advance();
+        }
+    }
+}
+
+void Player::Add_New_Shared_Special_Power_Timer(const SpecialPowerTemplate *tmplate, unsigned int timer)
+{
+    SpecialPowerReadyTimerType t;
+    t.m_id = tmplate->Get_ID();
+    t.m_frame = timer;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: m_frame, timer, correct names?

File: src/game/common/rts/money.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/common/rts/money.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:679]: Yes.

File: src/game/common/rts/player.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:2297]: No.

File: src/game/common/rts/player.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:2361]: No.

File: src/game/common/rts/player.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:2522]: This is correct.

File: src/game/common/rts/player.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:2740]: Yes.

File: src/game/common/rts/player.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: Fixed.

**Review by xezon — COMMENTED** [review]
Special Power shortcut button for Superweapons on the right side of the gui is not working properly. When multiple Superweapons are ready, it should always show them available, but here it will deactivate the button while some Superweapons are ready to fire.

shot_20230304_223345_3

**Review by xezon — APPROVED** [review]
Tested

* Building construction
* Eva events for Power Outage
* Eva events for Superweapons
* Cash Hack
* Home (H) button to find Command Center
* Special Power Shortcuts
* 1v1 vs Skirmish AI
* 3v3 with AI
* Surrender in Network match with Rocket Men in Bunker
* Shellmap
* Many units shooting on big test map

No unusual issues spotted that are introduced by this change. Looks ok.

**Pull request merged**